File: backend/src/main/java/com/mysite/knitly/domain/community/comment/service/CommentService.java
package com.mysite.knitly.domain.community.comment.service;

import com.mysite.knitly.domain.community.comment.dto.CommentCreateRequest;
import com.mysite.knitly.domain.community.comment.dto.CommentResponse;
import com.mysite.knitly.domain.community.comment.dto.CommentTreeResponse;
import com.mysite.knitly.domain.community.comment.dto.CommentUpdateRequest;
import com.mysite.knitly.domain.community.comment.entity.Comment;
import com.mysite.knitly.domain.community.comment.repository.CommentRepository;
import com.mysite.knitly.domain.community.post.entity.Post;
import com.mysite.knitly.domain.community.post.repository.PostRepository;
import com.mysite.knitly.domain.user.entity.User;
import com.mysite.knitly.global.exception.ErrorCode;
import com.mysite.knitly.global.exception.ServiceException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class CommentService {

    private final CommentRepository commentRepository;
    private final PostRepository postRepository;

    // 댓글 목록 - 전체 조회 후 서버에서 트리 구성(루트만 페이징)
    public Page<CommentTreeResponse> getComments(Long postId, String sort, int page, int size, User currentUser) {
        log.info("[CommentService] 댓글 목록 조회 요청 - postId={}, sort={}, page={}, size={}", postId, sort, page, size);

        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new ServiceException(ErrorCode.POST_NOT_FOUND));
        if (post.isDeleted()) {
            log.warn("[CommentService] 삭제된 게시글의 댓글 조회 시도 - postId={}", postId);
            throw new ServiceException(ErrorCode.POST_NOT_FOUND);
        }

        // 1) 해당 게시글의 모든 댓글을 한 번에 조회(등록순 정렬)
        List<Comment> all = commentRepository.findByPostIdAndDeletedFalseOrderByCreatedAtAsc(postId);

        // 2) parentId -> children 매핑
        Map<Long, List<Comment>> byParent = new HashMap<>();
        List<Comment> roots = all.stream()
                .filter(c -> c.getParent() == null)
                .collect(Collectors.toList());

        for (Comment c : all) {
            Long pid = (c.getParent() == null) ? null : c.getParent().getId();
            if (pid == null) continue;
            byParent.computeIfAbsent(pid, k -> new ArrayList<>()).add(c);
        }

        // children 정렬(등록순 기본)
        byParent.values().forEach(list -> list.sort(Comparator.comparing(Comment::getCreatedAt)));

        // 루트 정렬
        roots.sort(Comparator.comparing(Comment::getCreatedAt));
        if ("desc".equalsIgnoreCase(sort)) {
            Collections.reverse(roots);
            byParent.values().forEach(Collections::reverse);
        }

        // 3) 익명 넘버링 맵(게시글 작성자 제외)
        Map<Long, Integer> authorNoMap = buildAuthorNoMap(postId, post.getAuthor().getUserId());

        // 4) 루트 레벨만 페이징
        Pageable pageable = PageRequest.of(page, size);
        int from = Math.min(pageable.getPageNumber() * pageable.getPageSize(), roots.size());
        int to = Math.min(from + pageable.getPageSize(), roots.size());
        List<Comment> rootSlice = (from < to) ? roots.subList(from, to) : Collections.emptyList();

        // 5) 재귀 변환
        List<CommentTreeResponse> content = rootSlice.stream()
                .map(r -> toTreeRecursive(r, currentUser, authorNoMap, byParent))
                .collect(Collectors.toList());

        log.info("[CommentService] 댓글 목록 조회 완료 - postId={}, roots={}, returned={}", postId, roots.size(), content.size());
        return new PageImpl<>(content, pageable, roots.size());
    }

    // 댓글 개수
    public long count(Long postId) {
        long cnt = commentRepository.countByPostIdAndDeletedFalse(postId);
        log.info("[CommentService] 댓글 개수 조회 - postId={}, count={}", postId, cnt);
        return cnt;
    }

    // 댓글 작성
    @Transactional
    public CommentResponse create(CommentCreateRequest req, User currentUser) {
        log.info("[CommentService] 댓글 작성 요청 - postId={}, parentId={}", req.postId(), req.parentId());

        if (currentUser == null) {
            log.warn("[CommentService] 댓글 작성 실패 - 비인증 사용자");
            throw new ServiceException(ErrorCode.COMMENT_UNAUTHORIZED);
        }
        Post post = postRepository.findById(req.postId())
                .orElseThrow(() -> new ServiceException(ErrorCode.POST_NOT_FOUND));

        // parentId 검증
        Comment parent = null;
        if (req.parentId() != null) {
            parent = commentRepository.findById(req.parentId())
                    .orElseThrow(() -> new ServiceException(ErrorCode.COMMENT_NOT_FOUND));
            if (!parent.getPost().getId().equals(req.postId())) {
                log.warn("[CommentService] 댓글 작성 실패 - 서로 다른 게시글(parent/postId 불일치) parentId={}, postId={}",
                        req.parentId(), req.postId());
                throw new ServiceException(ErrorCode.BAD_REQUEST);
            }
        }

        // content trim & 공백만 입력 금지
        String trimmed = req.content() == null ? null : req.content().trim();
        if (trimmed == null || trimmed.isBlank()) {
            log.warn("[CommentService] 댓글 작성 실패 - 공백 또는 null 내용");
            throw new ServiceException(ErrorCode.BAD_REQUEST);
        }

        Comment saved = commentRepository.save(
                Comment.builder()
                        .post(post)
                        .author(currentUser) // 인증 사용자 그대로 사용
                        .content(trimmed)
                        .parent(parent)
                        .build()
        );

        Map<Long, Integer> authorNoMap = buildAuthorNoMap(req.postId(), post.getAuthor().getUserId());
        CommentResponse resp = toFlatResponse(saved, currentUser, authorNoMap);

        log.info("[CommentService] 댓글 작성 완료 - postId={}, commentId={}", req.postId(), saved.getId());
        return resp;
    }

    // 댓글 수정
    @Transactional
    public void update(Long commentId, CommentUpdateRequest req, User currentUser) {
        log.info("[CommentService] 댓글 수정 요청 - commentId={}", commentId);

        if (currentUser == null) {
            log.warn("[CommentService] 댓글 수정 실패 - 비인증 사용자");
            throw new ServiceException(ErrorCode.COMMENT_UNAUTHORIZED);
        }
        Comment c = commentRepository.findById(commentId)
                .orElseThrow(() -> new ServiceException(ErrorCode.COMMENT_NOT_FOUND));

        if (c.isDeleted()) {
            log.warn("[CommentService] 댓글 수정 실패 - 이미 삭제된 댓글 commentId={}", commentId);
            throw new ServiceException(ErrorCode.COMMENT_ALREADY_DELETED);
        }
        if (!c.isAuthor(currentUser)) {
            log.warn("[CommentService] 댓글 수정 실패 - 권한 없음 userId={}, authorId={}",
                    currentUser.getUserId(), c.getAuthor().getUserId());
            throw new ServiceException(ErrorCode.COMMENT_UPDATE_FORBIDDEN);
        }

        String trimmed = req.content() == null ? null : req.content().trim();
        if (trimmed == null || trimmed.isBlank()) {
            log.warn("[CommentService] 댓글 수정 실패 - 공백 또는 null 내용 commentId={}", commentId);
            throw new ServiceException(ErrorCode.BAD_REQUEST);
        }
        c.update(trimmed);
        log.info("[CommentService] 댓글 수정 완료 - commentId={}", commentId);
    }

    // 댓글 삭제
    @Transactional
    public void delete(Long commentId, User currentUser) {
        log.info("[CommentService] 댓글 삭제 요청 - commentId={}", commentId);

        if (currentUser == null) {
            log.warn("[CommentService] 댓글 삭제 실패 - 비인증 사용자");
            throw new ServiceException(ErrorCode.COMMENT_UNAUTHORIZED);
        }
        Comment c = commentRepository.findById(commentId)
                .orElseThrow(() -> new ServiceException(ErrorCode.COMMENT_NOT_FOUND));

        if (c.isDeleted()) {
            log.warn("[CommentService] 댓글 삭제 실패 - 이미 삭제된 댓글 commentId={}", commentId);
            throw new ServiceException(ErrorCode.COMMENT_ALREADY_DELETED);
        }
        if (!c.isAuthor(currentUser)) {
            log.warn("[CommentService] 댓글 삭제 실패 - 권한 없음 userId={}, authorId={}",
                    currentUser.getUserId(), c.getAuthor().getUserId());
            throw new ServiceException(ErrorCode.COMMENT_DELETE_FORBIDDEN);
        }
        c.softDelete();
        log.info("[CommentService] 댓글 삭제 완료 - commentId={}", commentId);
    }

    // 게시글 작성자는 번호 매핑에서 제외 (번호 없이 "익명의 털실")
    private Map<Long, Integer> buildAuthorNoMap(Long postId, Long postAuthorId) {
        List<Long> order = commentRepository.findAuthorOrderForPost(postId);
        Map<Long, Integer> map = new HashMap<>();
        int n = 1;
        for (Long uid : order) {
            if (uid == null) continue;
            if (uid.equals(postAuthorId)) continue; // 게시글 작성자 제외
            map.put(uid, n++);
        }
        return map;
    }

    // create 응답
    private CommentResponse toFlatResponse(Comment c, User currentUser, Map<Long, Integer> authorNoMap) {
        Long uid = (c.getAuthor() == null) ? null : c.getAuthor().getUserId();
        int no = (uid != null && authorNoMap.containsKey(uid)) ? authorNoMap.get(uid) : 0;
        String display = (no > 0) ? "익명의 털실 " + no : "익명의 털실";
        boolean mine = c.isAuthor(currentUser);

        return new CommentResponse(
                c.getId(),
                c.getContent(),
                uid,
                display,
                c.getCreatedAt(),
                mine
        );
    }

    // 재귀 트리 변환 (임의 깊이)
    private CommentTreeResponse toTreeRecursive(
            Comment node,
            User currentUser,
            Map<Long, Integer> authorNoMap,
            Map<Long, List<Comment>> byParent
    ) {
        List<Comment> children = byParent.getOrDefault(node.getId(), Collections.emptyList());

        List<CommentTreeResponse> childDtos = children.stream()
                .map(ch -> toTreeRecursive(ch, currentUser, authorNoMap, byParent))
                .collect(Collectors.toList());

        return new CommentTreeResponse(
                node.getId(),
                node.getContent(),
                node.getAuthor() == null ? null : node.getAuthor().getUserId(),
                displayName(node, authorNoMap),
                node.getCreatedAt(),
                isMine(node, currentUser),
                node.getParent() == null ? null : node.getParent().getId(),
                childDtos
        );
    }

    private boolean isMine(Comment c, User currentUser) {
        return c.isAuthor(currentUser);
    }

    private String displayName(Comment c, Map<Long, Integer> authorNoMap) {
        Long uid = (c.getAuthor() == null) ? null : c.getAuthor().getUserId();
        int no = (uid != null && authorNoMap.containsKey(uid)) ? authorNoMap.get(uid) : 0;
        return (no > 0) ? "익명의 털실 " + no : "익명의 털실";
    }
}

File: backend/src/main/kotlin/com/mysite/knitly/domain/community/comment/service/CommentService.kt
package com.mysite.knitly.domain.community.comment.service

import com.mysite.knitly.domain.community.comment.dto.CommentCreateRequest
import com.mysite.knitly.domain.community.comment.dto.CommentResponse
import com.mysite.knitly.domain.community.comment.dto.CommentTreeResponse
import com.mysite.knitly.domain.community.comment.dto.CommentUpdateRequest
import com.mysite.knitly.domain.community.comment.entity.Comment
import com.mysite.knitly.domain.community.comment.repository.CommentRepository
import com.mysite.knitly.domain.community.post.entity.Post
import com.mysite.knitly.domain.community.post.repository.PostRepository
import com.mysite.knitly.domain.user.entity.User
import com.mysite.knitly.global.exception.ErrorCode
import com.mysite.knitly.global.exception.ServiceException
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class CommentService(
    private val commentRepository: CommentRepository,
    private val postRepository: PostRepository
) {

    private val logger = LoggerFactory.getLogger(CommentService::class.java)

    fun getComments(postId: Long, sort: String, page: Int, size: Int, currentUser: User?): Page<CommentTreeResponse> {
        logger.info("[CommentService] 댓글 목록 조회 요청 - postId={}, sort={}, page={}, size={}", postId, sort, page, size)

        val post: Post = postRepository.findById(postId)
            .orElseThrow { ServiceException(ErrorCode.POST_NOT_FOUND) }

        if (post.deleted) {
            logger.warn("[CommentService] 삭제된 게시글의 댓글 조회 시도 - postId={}", postId)
            throw ServiceException(ErrorCode.POST_NOT_FOUND)
        }

        val all: List<Comment> = commentRepository.findByPostIdAndDeletedFalseOrderByCreatedAtAsc(postId)

        val byParent: MutableMap<Long, MutableList<Comment>> = HashMap()
        val roots: MutableList<Comment> = all
            .filter { it.parent == null }
            .toMutableList()

        for (c in all) {
            val pid = c.parent?.id
            if (pid == null) continue
            byParent.computeIfAbsent(pid) { mutableListOf() }.add(c)
        }

        byParent.values.forEach { list ->
            list.sortBy { it.createdAt }
        }

        roots.sortBy { it.createdAt }
        if (sort.equals("desc", ignoreCase = true)) {
            roots.reverse()
            byParent.values.forEach { it.reverse() }
        }

        val authorNoMap = buildAuthorNoMap(postId, post.author.userId)

        val pageable: Pageable = PageRequest.of(page, size)
        val from = (pageable.pageNumber * pageable.pageSize).coerceAtMost(roots.size)
        val to = (from + pageable.pageSize).coerceAtMost(roots.size)
        val rootSlice: List<Comment> =
            if (from < to) roots.subList(from, to) else emptyList()

        val content: List<CommentTreeResponse> = rootSlice
            .map { r -> toTreeRecursive(r, currentUser, authorNoMap, byParent) }

        logger.info(
            "[CommentService] 댓글 목록 조회 완료 - postId={}, roots={}, returned={}",
            postId,
            roots.size,
            content.size
        )

        return PageImpl(content, pageable, roots.size.toLong())
    }

    fun count(postId: Long): Long {
        val cnt = commentRepository.countByPostIdAndDeletedFalse(postId)
        logger.info("[CommentService] 댓글 개수 조회 - postId={}, count={}", postId, cnt)
        return cnt
    }

    @Transactional
    fun create(req: CommentCreateRequest, currentUser: User?): CommentResponse {
        logger.info("[CommentService] 댓글 작성 요청 - postId={}, parentId={}", req.postId, req.parentId)

        if (currentUser == null) {
            logger.warn("[CommentService] 댓글 작성 실패 - 비인증 사용자")
            throw ServiceException(ErrorCode.COMMENT_UNAUTHORIZED)
        }

        val post: Post = postRepository.findById(req.postId)
            .orElseThrow { ServiceException(ErrorCode.POST_NOT_FOUND) }

        var parent: Comment? = null
        val parentId = req.parentId
        if (parentId != null) {
            parent = commentRepository.findById(parentId)
                .orElseThrow { ServiceException(ErrorCode.COMMENT_NOT_FOUND) }

            if (parent.post.id != req.postId) {
                logger.warn(
                    "[CommentService] 댓글 작성 실패 - 서로 다른 게시글(parent/postId 불일치) parentId={}, postId={}",
                    parentId,
                    req.postId
                )
                throw ServiceException(ErrorCode.BAD_REQUEST)
            }
        }

        val trimmed = req.content.trim()
        if (trimmed.isBlank()) {
            logger.warn("[CommentService] 댓글 작성 실패 - 공백 또는 null 내용")
            throw ServiceException(ErrorCode.BAD_REQUEST)
        }

        val saved = commentRepository.save(
            Comment(
                content = trimmed,
                author = currentUser,
                post = post,
                parent = parent
            )
        )

        val authorNoMap = buildAuthorNoMap(req.postId, post.author.userId)
        val resp = toFlatResponse(saved, currentUser, authorNoMap)

        logger.info("[CommentService] 댓글 작성 완료 - postId={}, commentId={}", req.postId, saved.id)
        return resp
    }

    @Transactional
    fun update(commentId: Long, req: CommentUpdateRequest, currentUser: User?) {
        logger.info("[CommentService] 댓글 수정 요청 - commentId={}", commentId)

        if (currentUser == null) {
            logger.warn("[CommentService] 댓글 수정 실패 - 비인증 사용자")
            throw ServiceException(ErrorCode.COMMENT_UNAUTHORIZED)
        }

        val c: Comment = commentRepository.findById(commentId)
            .orElseThrow { ServiceException(ErrorCode.COMMENT_NOT_FOUND) }

        if (c.deleted) {
            logger.warn("[CommentService] 댓글 수정 실패 - 이미 삭제된 댓글 commentId={}", commentId)
            throw ServiceException(ErrorCode.COMMENT_ALREADY_DELETED)
        }

        if (!c.isAuthor(currentUser)) {
            logger.warn(
                "[CommentService] 댓글 수정 실패 - 권한 없음 userId={}, authorId={}",
                currentUser.userId,
                c.author.userId
            )
            throw ServiceException(ErrorCode.COMMENT_UPDATE_FORBIDDEN)
        }

        val trimmed = req.content.trim()
        if (trimmed.isBlank()) {
            logger.warn("[CommentService] 댓글 수정 실패 - 공백 또는 null 내용 commentId={}", commentId)
            throw ServiceException(ErrorCode.BAD_REQUEST)
        }

        c.update(trimmed)
        logger.info("[CommentService] 댓글 수정 완료 - commentId={}", commentId)
    }

    @Transactional
    fun delete(commentId: Long, currentUser: User?) {
        logger.info("[CommentService] 댓글 삭제 요청 - commentId={}", commentId)

        if (currentUser == null) {
            logger.warn("[CommentService] 댓글 삭제 실패 - 비인증 사용자")
            throw ServiceException(ErrorCode.COMMENT_UNAUTHORIZED)
        }

        val c: Comment = commentRepository.findById(commentId)
            .orElseThrow { ServiceException(ErrorCode.COMMENT_NOT_FOUND) }

        if (c.deleted) {
            logger.warn("[CommentService] 댓글 삭제 실패 - 이미 삭제된 댓글 commentId={}", commentId)
            throw ServiceException(ErrorCode.COMMENT_ALREADY_DELETED)
        }

        if (!c.isAuthor(currentUser)) {
            logger.warn(
                "[CommentService] 댓글 삭제 실패 - 권한 없음 userId={}, authorId={}",
                currentUser.userId,
                c.author.userId
            )
            throw ServiceException(ErrorCode.COMMENT_DELETE_FORBIDDEN)
        }

        c.softDelete()
        logger.info("[CommentService] 댓글 삭제 완료 - commentId={}", commentId)
    }

    private fun buildAuthorNoMap(postId: Long, postAuthorId: Long): Map<Long, Int> {
        val order: List<Long> = commentRepository.findAuthorOrderForPost(postId)
        val map: MutableMap<Long, Int> = HashMap()
        var n = 1
        for (uid in order) {
            if (uid == null) continue
            if (uid == postAuthorId) continue
            map[uid] = n++
        }
        return map
    }

    private fun toFlatResponse(
        c: Comment,
        currentUser: User?,
        authorNoMap: Map<Long, Int>
    ): CommentResponse {
        val commentId = c.id ?: throw IllegalStateException("Comment id is null")
        val uid = c.author.userId
        val no = authorNoMap[uid] ?: 0
        val display = if (no > 0) "익명의 털실 $no" else "익명의 털실"
        val mine = c.isAuthor(currentUser)

        return CommentResponse(
            commentId,
            c.content,
            uid,
            display,
            c.createdAt,
            mine
        )
    }

    private fun toTreeRecursive(
        node: Comment,
        currentUser: User?,
        authorNoMap: Map<Long, Int>,
        byParent: Map<Long, List<Comment>>
    ): CommentTreeResponse {
        val nodeId = node.id ?: throw IllegalStateException("Comment id is null")
        val children: List<Comment> = byParent[nodeId] ?: emptyList()

        val childDtos: List<CommentTreeResponse> = children
            .map { ch -> toTreeRecursive(ch, currentUser, authorNoMap, byParent) }

        return CommentTreeResponse(
            nodeId,
            node.content,
            node.author.userId,
            displayName(node, authorNoMap),
            node.createdAt,
            isMine(node, currentUser),
            node.parent?.id,
            childDtos
        )
    }

    private fun isMine(c: Comment, currentUser: User?): Boolean {
        return c.isAuthor(currentUser)
    }

    private fun displayName(c: Comment, authorNoMap: Map<Long, Int>): String {
        val uid = c.author.userId
        val no = authorNoMap[uid] ?: 0
        return if (no > 0) "익명의 털실 $no" else "익명의 털실"
    }
}
